File: Resources/OpenMP_hello_world-20240125/program_hello_v2.f90
! a little more sophisticated version of hello world
module data
#ifdef _OPENMP
  use OMP_LIB
#endif
  implicit none

contains

  subroutine hello()
    implicit none
    integer :: my_rank, thread_count

#ifdef _OPENMP
    my_rank = OMP_GET_THREAD_NUM()
    thread_count = OMP_GET_NUM_THREADS()
#else
    my_rank = 0
    thread_count = 1 
#endif

    write(*, *) 'Hello from thread', my_rank, 'out of total threads of', thread_count

    !$OMP PARALLEL NUM_THREADS(4)
    write(*, *) 'Hi'
    !$OMP END PARALLEL

  end subroutine hello
  
end module data

program main
  use data
  implicit none
  integer :: thread_count = 4 
  character(100) :: numchar
  character(100) :: name

  if (COMMAND_ARGUMENT_COUNT() .ne. 1) then
     write(*, *) 'Error, one command line arguments is required, stopping the program'
     STOP
  end if

  

  call GET_COMMAND_ARGUMENT(0,name)
  write(*, *) name
  
  call GET_COMMAND_ARGUMENT(1, numchar)
  READ(numchar, *) thread_count

  !  call OMP_SET_MAX_ACTIVE_LEVELS(3)
  call OMP_SET_NESTED(.false.)
  
  !$OMP PARALLEL NUM_THREADS(thread_count)
  call hello()
  !$OMP END PARALLEL
    

end program main

File: Resources/OpenMP_hello_world-20240125/program_hello_v2.c
/* A little more sophisticated version of hello world */
#include<stdio.h>
#include<stdlib.h>
#ifdef _OPENMP
#include<omp.h>
#endif

void Hello(void)
{
#ifdef _OPENMP
  int my_rank = omp_get_thread_num();
  int thread_count = omp_get_num_threads();
#else
  int my_rank = 0;
  int thread_count = 1;
#endif

  printf("\n Hello from thread %d out of total threads of %d", my_rank, thread_count);
}

int main(int argc, char* argv[])
{
  int thread_count = 1;

  if (argc == 2)
    {
      thread_count = strtol(argv[1], NULL, 10);
    }
  else
    {
      printf("\n A command line argument other than name of the executable is required...exiting the program..\n");
      return 1;
    }
  
#pragma omp parallel num_threads(thread_count)
  Hello();

  return 0;
}
